// @vectorize-io/hindsight-client/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  HindsightClient: () => HindsightClient,
  HindsightError: () => HindsightError,
  createClient: () => createClient,
  createConfig: () => createConfig,
  sdk: () => sdk_gen_exports
});
module.exports = __toCommonJS(index_exports);

// generated/core/bodySerializer.gen.ts
var serializeFormDataPair = (data, key, value) => {
  if (typeof value === "string" || value instanceof Blob) {
    data.append(key, value);
  } else if (value instanceof Date) {
    data.append(key, value.toISOString());
  } else {
    data.append(key, JSON.stringify(value));
  }
};
var formDataBodySerializer = {
  bodySerializer: (body) => {
    const data = new FormData();
    Object.entries(body).forEach(([key, value]) => {
      if (value === void 0 || value === null) {
        return;
      }
      if (Array.isArray(value)) {
        value.forEach((v) => serializeFormDataPair(data, key, v));
      } else {
        serializeFormDataPair(data, key, value);
      }
    });
    return data;
  }
};
var jsonBodySerializer = {
  bodySerializer: (body) => JSON.stringify(
    body,
    (_key, value) => typeof value === "bigint" ? value.toString() : value
  )
};

// generated/core/params.gen.ts
var extraPrefixesMap = {
  $body_: "body",
  $headers_: "headers",
  $path_: "path",
  $query_: "query"
};
var extraPrefixes = Object.entries(extraPrefixesMap);

// generated/core/serverSentEvents.gen.ts
var createSseClient = ({
  onRequest,
  onSseError,
  onSseEvent,
  responseTransformer,
  responseValidator,
  sseDefaultRetryDelay,
  sseMaxRetryAttempts,
  sseMaxRetryDelay,
  sseSleepFn,
  url,
  ...options
}) => {
  let lastEventId;
  const sleep = sseSleepFn ?? ((ms) => new Promise((resolve) => setTimeout(resolve, ms)));
  const createStream = async function* () {
    let retryDelay = sseDefaultRetryDelay ?? 3e3;
    let attempt = 0;
    const signal = options.signal ?? new AbortController().signal;
    while (true) {
      if (signal.aborted) break;
      attempt++;
      const headers = options.headers instanceof Headers ? options.headers : new Headers(options.headers);
      if (lastEventId !== void 0) {
        headers.set("Last-Event-ID", lastEventId);
      }
      try {
        const requestInit = {
          redirect: "follow",
          ...options,
          body: options.serializedBody,
          headers,
          signal
        };
        let request = new Request(url, requestInit);
        if (onRequest) {
          request = await onRequest(url, requestInit);
        }
        const _fetch = options.fetch ?? globalThis.fetch;
        const response = await _fetch(request);
        if (!response.ok)
          throw new Error(
            `SSE failed: ${response.status} ${response.statusText}`
          );
        if (!response.body) throw new Error("No body in SSE response");
        const reader = response.body.pipeThrough(new TextDecoderStream()).getReader();
        let buffer = "";
        const abortHandler = () => {
          try {
            reader.cancel();
          } catch {
          }
        };
        signal.addEventListener("abort", abortHandler);
        try {
          while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            buffer += value;
            const chunks = buffer.split("\n\n");
            buffer = chunks.pop() ?? "";
            for (const chunk of chunks) {
              const lines = chunk.split("\n");
              const dataLines = [];
              let eventName;
              for (const line of lines) {
                if (line.startsWith("data:")) {
                  dataLines.push(line.replace(/^data:\s*/, ""));
                } else if (line.startsWith("event:")) {
                  eventName = line.replace(/^event:\s*/, "");
                } else if (line.startsWith("id:")) {
                  lastEventId = line.replace(/^id:\s*/, "");
                } else if (line.startsWith("retry:")) {
                  const parsed = Number.parseInt(
                    line.replace(/^retry:\s*/, ""),
                    10
                  );
                  if (!Number.isNaN(parsed)) {
                    retryDelay = parsed;
                  }
                }
              }
              let data;
              let parsedJson = false;
              if (dataLines.length) {
                const rawData = dataLines.join("\n");
                try {
                  data = JSON.parse(rawData);
                  parsedJson = true;
                } catch {
                  data = rawData;
                }
              }
              if (parsedJson) {
                if (responseValidator) {
                  await responseValidator(data);
                }
                if (responseTransformer) {
                  data = await responseTransformer(data);
                }
              }
              onSseEvent?.({
                data,
                event: eventName,
                id: lastEventId,
                retry: retryDelay
              });
              if (dataLines.length) {
                yield data;
              }
            }
          }
        } finally {
          signal.removeEventListener("abort", abortHandler);
          reader.releaseLock();
        }
        break;
      } catch (error) {
        onSseError?.(error);
        if (sseMaxRetryAttempts !== void 0 && attempt >= sseMaxRetryAttempts) {
          break;
        }
        const backoff = Math.min(
          retryDelay * 2 ** (attempt - 1),
          sseMaxRetryDelay ?? 3e4
        );
        await sleep(backoff);
      }
    }
  };
  const stream = createStream();
  return { stream };
};

// generated/core/pathSerializer.gen.ts
var separatorArrayExplode = (style) => {
  switch (style) {
    case "label":
      return ".";
    case "matrix":
      return ";";
    case "simple":
      return ",";
    default:
      return "&";
  }
};
var separatorArrayNoExplode = (style) => {
  switch (style) {
    case "form":
      return ",";
    case "pipeDelimited":
      return "|";
    case "spaceDelimited":
      return "%20";
    default:
      return ",";
  }
};
var separatorObjectExplode = (style) => {
  switch (style) {
    case "label":
      return ".";
    case "matrix":
      return ";";
    case "simple":
      return ",";
    default:
      return "&";
  }
};
var serializeArrayParam = ({
  allowReserved,
  explode,
  name,
  style,
  value
}) => {
  if (!explode) {
    const joinedValues2 = (allowReserved ? value : value.map((v) => encodeURIComponent(v))).join(separatorArrayNoExplode(style));
    switch (style) {
      case "label":
        return `.${joinedValues2}`;
      case "matrix":
        return `;${name}=${joinedValues2}`;
      case "simple":
        return joinedValues2;
      default:
        return `${name}=${joinedValues2}`;
    }
  }
  const separator = separatorArrayExplode(style);
  const joinedValues = value.map((v) => {
    if (style === "label" || style === "simple") {
      return allowReserved ? v : encodeURIComponent(v);
    }
    return serializePrimitiveParam({
      allowReserved,
      name,
      value: v
    });
  }).join(separator);
  return style === "label" || style === "matrix" ? separator + joinedValues : joinedValues;
};
var serializePrimitiveParam = ({
  allowReserved,
  name,
  value
}) => {
  if (value === void 0 || value === null) {
    return "";
  }
  if (typeof value === "object") {
    throw new Error(
      "Deeply-nested arrays/objects aren\u2019t supported. Provide your own `querySerializer()` to handle these."
    );
  }
  return `${name}=${allowReserved ? value : encodeURIComponent(value)}`;
};
var serializeObjectParam = ({
  allowReserved,
  explode,
  name,
  style,
  value,
  valueOnly
}) => {
  if (value instanceof Date) {
    return valueOnly ? value.toISOString() : `${name}=${value.toISOString()}`;
  }
  if (style !== "deepObject" && !explode) {
    let values = [];
    Object.entries(value).forEach(([key, v]) => {
      values = [
        ...values,
        key,
        allowReserved ? v : encodeURIComponent(v)
      ];
    });
    const joinedValues2 = values.join(",");
    switch (style) {
      case "form":
        return `${name}=${joinedValues2}`;
      case "label":
        return `.${joinedValues2}`;
      case "matrix":
        return `;${name}=${joinedValues2}`;
      default:
        return joinedValues2;
    }
  }
  const separator = separatorObjectExplode(style);
  const joinedValues = Object.entries(value).map(
    ([key, v]) => serializePrimitiveParam({
      allowReserved,
      name: style === "deepObject" ? `${name}[${key}]` : key,
      value: v
    })
  ).join(separator);
  return style === "label" || style === "matrix" ? separator + joinedValues : joinedValues;
};

// generated/core/utils.gen.ts
var PATH_PARAM_RE = /\{[^{}]+\}/g;
var defaultPathSerializer = ({ path, url: _url }) => {
  let url = _url;
  const matches = _url.match(PATH_PARAM_RE);
  if (matches) {
    for (const match of matches) {
      let explode = false;
      let name = match.substring(1, match.length - 1);
      let style = "simple";
      if (name.endsWith("*")) {
        explode = true;
        name = name.substring(0, name.length - 1);
      }
      if (name.startsWith(".")) {
        name = name.substring(1);
        style = "label";
      } else if (name.startsWith(";")) {
        name = name.substring(1);
        style = "matrix";
      }
      const value = path[name];
      if (value === void 0 || value === null) {
        continue;
      }
      if (Array.isArray(value)) {
        url = url.replace(
          match,
          serializeArrayParam({ explode, name, style, value })
        );
        continue;
      }
      if (typeof value === "object") {
        url = url.replace(
          match,
          serializeObjectParam({
            explode,
            name,
            style,
            value,
            valueOnly: true
          })
        );
        continue;
      }
      if (style === "matrix") {
        url = url.replace(
          match,
          `;${serializePrimitiveParam({
            name,
            value
          })}`
        );
        continue;
      }
      const replaceValue = encodeURIComponent(
        style === "label" ? `.${value}` : value
      );
      url = url.replace(match, replaceValue);
    }
  }
  return url;
};
var getUrl = ({
  baseUrl,
  path,
  query,
  querySerializer,
  url: _url
}) => {
  const pathUrl = _url.startsWith("/") ? _url : `/${_url}`;
  let url = (baseUrl ?? "") + pathUrl;
  if (path) {
    url = defaultPathSerializer({ path, url });
  }
  let search = query ? querySerializer(query) : "";
  if (search.startsWith("?")) {
    search = search.substring(1);
  }
  if (search) {
    url += `?${search}`;
  }
  return url;
};
function getValidRequestBody(options) {
  const hasBody = options.body !== void 0;
  const isSerializedBody = hasBody && options.bodySerializer;
  if (isSerializedBody) {
    if ("serializedBody" in options) {
      const hasSerializedBody = options.serializedBody !== void 0 && options.serializedBody !== "";
      return hasSerializedBody ? options.serializedBody : null;
    }
    return options.body !== "" ? options.body : null;
  }
  if (hasBody) {
    return options.body;
  }
  return void 0;
}

// generated/core/auth.gen.ts
var getAuthToken = async (auth, callback) => {
  const token = typeof callback === "function" ? await callback(auth) : callback;
  if (!token) {
    return;
  }
  if (auth.scheme === "bearer") {
    return `Bearer ${token}`;
  }
  if (auth.scheme === "basic") {
    return `Basic ${btoa(token)}`;
  }
  return token;
};

// generated/client/utils.gen.ts
var createQuerySerializer = ({
  parameters = {},
  ...args
} = {}) => {
  const querySerializer = (queryParams) => {
    const search = [];
    if (queryParams && typeof queryParams === "object") {
      for (const name in queryParams) {
        const value = queryParams[name];
        if (value === void 0 || value === null) {
          continue;
        }
        const options = parameters[name] || args;
        if (Array.isArray(value)) {
          const serializedArray = serializeArrayParam({
            allowReserved: options.allowReserved,
            explode: true,
            name,
            style: "form",
            value,
            ...options.array
          });
          if (serializedArray) search.push(serializedArray);
        } else if (typeof value === "object") {
          const serializedObject = serializeObjectParam({
            allowReserved: options.allowReserved,
            explode: true,
            name,
            style: "deepObject",
            value,
            ...options.object
          });
          if (serializedObject) search.push(serializedObject);
        } else {
          const serializedPrimitive = serializePrimitiveParam({
            allowReserved: options.allowReserved,
            name,
            value
          });
          if (serializedPrimitive) search.push(serializedPrimitive);
        }
      }
    }
    return search.join("&");
  };
  return querySerializer;
};
var getParseAs = (contentType) => {
  if (!contentType) {
    return "stream";
  }
  const cleanContent = contentType.split(";")[0]?.trim();
  if (!cleanContent) {
    return;
  }
  if (cleanContent.startsWith("application/json") || cleanContent.endsWith("+json")) {
    return "json";
  }
  if (cleanContent === "multipart/form-data") {
    return "formData";
  }
  if (["application/", "audio/", "image/", "video/"].some(
    (type) => cleanContent.startsWith(type)
  )) {
    return "blob";
  }
  if (cleanContent.startsWith("text/")) {
    return "text";
  }
  return;
};
var checkForExistence = (options, name) => {
  if (!name) {
    return false;
  }
  if (options.headers.has(name) || options.query?.[name] || options.headers.get("Cookie")?.includes(`${name}=`)) {
    return true;
  }
  return false;
};
var setAuthParams = async ({
  security,
  ...options
}) => {
  for (const auth of security) {
    if (checkForExistence(options, auth.name)) {
      continue;
    }
    const token = await getAuthToken(auth, options.auth);
    if (!token) {
      continue;
    }
    const name = auth.name ?? "Authorization";
    switch (auth.in) {
      case "query":
        if (!options.query) {
          options.query = {};
        }
        options.query[name] = token;
        break;
      case "cookie":
        options.headers.append("Cookie", `${name}=${token}`);
        break;
      case "header":
      default:
        options.headers.set(name, token);
        break;
    }
  }
};
var buildUrl = (options) => getUrl({
  baseUrl: options.baseUrl,
  path: options.path,
  query: options.query,
  querySerializer: typeof options.querySerializer === "function" ? options.querySerializer : createQuerySerializer(options.querySerializer),
  url: options.url
});
var mergeConfigs = (a, b) => {
  const config = { ...a, ...b };
  if (config.baseUrl?.endsWith("/")) {
    config.baseUrl = config.baseUrl.substring(0, config.baseUrl.length - 1);
  }
  config.headers = mergeHeaders(a.headers, b.headers);
  return config;
};
var headersEntries = (headers) => {
  const entries = [];
  headers.forEach((value, key) => {
    entries.push([key, value]);
  });
  return entries;
};
var mergeHeaders = (...headers) => {
  const mergedHeaders = new Headers();
  for (const header of headers) {
    if (!header) {
      continue;
    }
    const iterator = header instanceof Headers ? headersEntries(header) : Object.entries(header);
    for (const [key, value] of iterator) {
      if (value === null) {
        mergedHeaders.delete(key);
      } else if (Array.isArray(value)) {
        for (const v of value) {
          mergedHeaders.append(key, v);
        }
      } else if (value !== void 0) {
        mergedHeaders.set(
          key,
          typeof value === "object" ? JSON.stringify(value) : value
        );
      }
    }
  }
  return mergedHeaders;
};
var Interceptors = class {
  constructor() {
    this.fns = [];
  }
  clear() {
    this.fns = [];
  }
  eject(id) {
    const index = this.getInterceptorIndex(id);
    if (this.fns[index]) {
      this.fns[index] = null;
    }
  }
  exists(id) {
    const index = this.getInterceptorIndex(id);
    return Boolean(this.fns[index]);
  }
  getInterceptorIndex(id) {
    if (typeof id === "number") {
      return this.fns[id] ? id : -1;
    }
    return this.fns.indexOf(id);
  }
  update(id, fn) {
    const index = this.getInterceptorIndex(id);
    if (this.fns[index]) {
      this.fns[index] = fn;
      return id;
    }
    return false;
  }
  use(fn) {
    this.fns.push(fn);
    return this.fns.length - 1;
  }
};
var createInterceptors = () => ({
  error: new Interceptors(),
  request: new Interceptors(),
  response: new Interceptors()
});
var defaultQuerySerializer = createQuerySerializer({
  allowReserved: false,
  array: {
    explode: true,
    style: "form"
  },
  object: {
    explode: true,
    style: "deepObject"
  }
});
var defaultHeaders = {
  "Content-Type": "application/json"
};
var createConfig = (override = {}) => ({
  ...jsonBodySerializer,
  headers: defaultHeaders,
  parseAs: "auto",
  querySerializer: defaultQuerySerializer,
  ...override
});

// generated/client/client.gen.ts
var createClient = (config = {}) => {
  let _config = mergeConfigs(createConfig(), config);
  const getConfig = () => ({ ..._config });
  const setConfig = (config2) => {
    _config = mergeConfigs(_config, config2);
    return getConfig();
  };
  const interceptors = createInterceptors();
  const beforeRequest = async (options) => {
    const opts = {
      ..._config,
      ...options,
      fetch: options.fetch ?? _config.fetch ?? globalThis.fetch,
      headers: mergeHeaders(_config.headers, options.headers),
      serializedBody: void 0
    };
    if (opts.security) {
      await setAuthParams({
        ...opts,
        security: opts.security
      });
    }
    if (opts.requestValidator) {
      await opts.requestValidator(opts);
    }
    if (opts.body !== void 0 && opts.bodySerializer) {
      opts.serializedBody = opts.bodySerializer(opts.body);
    }
    if (opts.body === void 0 || opts.serializedBody === "") {
      opts.headers.delete("Content-Type");
    }
    const url = buildUrl(opts);
    return { opts, url };
  };
  const request = async (options) => {
    const { opts, url } = await beforeRequest(options);
    const { client: _client, ...optsForRequest } = opts;
    const requestInit = {
      redirect: "follow",
      ...optsForRequest,
      body: getValidRequestBody(opts)
    };
    let request2 = new Request(url, requestInit);
    for (const fn of interceptors.request.fns) {
      if (fn) {
        request2 = await fn(request2, opts);
      }
    }
    const _fetch = opts.fetch;
    let response;
    try {
      response = await _fetch(request2);
    } catch (error2) {
      let finalError2 = error2;
      for (const fn of interceptors.error.fns) {
        if (fn) {
          finalError2 = await fn(
            error2,
            void 0,
            request2,
            opts
          );
        }
      }
      finalError2 = finalError2 || {};
      if (opts.throwOnError) {
        throw finalError2;
      }
      return opts.responseStyle === "data" ? void 0 : {
        error: finalError2,
        request: request2,
        response: void 0
      };
    }
    for (const fn of interceptors.response.fns) {
      if (fn) {
        response = await fn(response, request2, opts);
      }
    }
    const result = {
      request: request2,
      response
    };
    if (response.ok) {
      const parseAs = (opts.parseAs === "auto" ? getParseAs(response.headers.get("Content-Type")) : opts.parseAs) ?? "json";
      if (response.status === 204 || response.headers.get("Content-Length") === "0") {
        let emptyData;
        switch (parseAs) {
          case "arrayBuffer":
          case "blob":
          case "text":
            emptyData = await response[parseAs]();
            break;
          case "formData":
            emptyData = new FormData();
            break;
          case "stream":
            emptyData = response.body;
            break;
          case "json":
          default:
            emptyData = {};
            break;
        }
        return opts.responseStyle === "data" ? emptyData : {
          data: emptyData,
          ...result
        };
      }
      let data;
      switch (parseAs) {
        case "arrayBuffer":
        case "blob":
        case "formData":
        case "json":
        case "text":
          data = await response[parseAs]();
          break;
        case "stream":
          return opts.responseStyle === "data" ? response.body : {
            data: response.body,
            ...result
          };
      }
      if (parseAs === "json") {
        if (opts.responseValidator) {
          await opts.responseValidator(data);
        }
        if (opts.responseTransformer) {
          data = await opts.responseTransformer(data);
        }
      }
      return opts.responseStyle === "data" ? data : {
        data,
        ...result
      };
    }
    const textError = await response.text();
    let jsonError;
    try {
      jsonError = JSON.parse(textError);
    } catch {
    }
    const error = jsonError ?? textError;
    let finalError = error;
    for (const fn of interceptors.error.fns) {
      if (fn) {
        finalError = await fn(error, response, request2, opts);
      }
    }
    finalError = finalError || {};
    if (opts.throwOnError) {
      throw finalError;
    }
    return opts.responseStyle === "data" ? void 0 : {
      error: finalError,
      ...result
    };
  };
  const makeMethodFn = (method) => (options) => request({ ...options, method });
  const makeSseFn = (method) => async (options) => {
    const { opts, url } = await beforeRequest(options);
    return createSseClient({
      ...opts,
      body: opts.body,
      headers: opts.headers,
      method,
      onRequest: async (url2, init) => {
        let request2 = new Request(url2, init);
        for (const fn of interceptors.request.fns) {
          if (fn) {
            request2 = await fn(request2, opts);
          }
        }
        return request2;
      },
      url
    });
  };
  return {
    buildUrl,
    connect: makeMethodFn("CONNECT"),
    delete: makeMethodFn("DELETE"),
    get: makeMethodFn("GET"),
    getConfig,
    head: makeMethodFn("HEAD"),
    interceptors,
    options: makeMethodFn("OPTIONS"),
    patch: makeMethodFn("PATCH"),
    post: makeMethodFn("POST"),
    put: makeMethodFn("PUT"),
    request,
    setConfig,
    sse: {
      connect: makeSseFn("CONNECT"),
      delete: makeSseFn("DELETE"),
      get: makeSseFn("GET"),
      head: makeSseFn("HEAD"),
      options: makeSseFn("OPTIONS"),
      patch: makeSseFn("PATCH"),
      post: makeSseFn("POST"),
      put: makeSseFn("PUT"),
      trace: makeSseFn("TRACE")
    },
    trace: makeMethodFn("TRACE")
  };
};

// generated/sdk.gen.ts
var sdk_gen_exports = {};
__export(sdk_gen_exports, {
  addBankBackground: () => addBankBackground,
  auditLogStats: () => auditLogStats,
  cancelOperation: () => cancelOperation,
  clearBankMemories: () => clearBankMemories,
  clearMemoryObservations: () => clearMemoryObservations,
  clearObservations: () => clearObservations,
  createDirective: () => createDirective,
  createMentalModel: () => createMentalModel,
  createOrUpdateBank: () => createOrUpdateBank,
  createWebhook: () => createWebhook,
  deleteBank: () => deleteBank,
  deleteDirective: () => deleteDirective,
  deleteDocument: () => deleteDocument,
  deleteMentalModel: () => deleteMentalModel,
  deleteWebhook: () => deleteWebhook,
  fileRetain: () => fileRetain,
  getAgentStats: () => getAgentStats,
  getBankConfig: () => getBankConfig,
  getBankProfile: () => getBankProfile,
  getChunk: () => getChunk,
  getDirective: () => getDirective,
  getDocument: () => getDocument,
  getEntity: () => getEntity,
  getGraph: () => getGraph,
  getMemory: () => getMemory,
  getMentalModel: () => getMentalModel,
  getMentalModelHistory: () => getMentalModelHistory,
  getObservationHistory: () => getObservationHistory,
  getOperationStatus: () => getOperationStatus,
  getVersion: () => getVersion,
  healthEndpointHealthGet: () => healthEndpointHealthGet,
  listAuditLogs: () => listAuditLogs,
  listBanks: () => listBanks,
  listDirectives: () => listDirectives,
  listDocuments: () => listDocuments,
  listEntities: () => listEntities,
  listMemories: () => listMemories,
  listMentalModels: () => listMentalModels,
  listOperations: () => listOperations,
  listTags: () => listTags,
  listWebhookDeliveries: () => listWebhookDeliveries,
  listWebhooks: () => listWebhooks,
  metricsEndpointMetricsGet: () => metricsEndpointMetricsGet,
  recallMemories: () => recallMemories,
  recoverConsolidation: () => recoverConsolidation,
  reflect: () => reflect,
  refreshMentalModel: () => refreshMentalModel,
  regenerateEntityObservations: () => regenerateEntityObservations,
  resetBankConfig: () => resetBankConfig,
  retainMemories: () => retainMemories,
  retryOperation: () => retryOperation,
  triggerConsolidation: () => triggerConsolidation,
  updateBank: () => updateBank,
  updateBankConfig: () => updateBankConfig,
  updateBankDisposition: () => updateBankDisposition,
  updateDirective: () => updateDirective,
  updateDocument: () => updateDocument,
  updateMentalModel: () => updateMentalModel,
  updateWebhook: () => updateWebhook
});

// generated/client.gen.ts
var client = createClient(createConfig());

// generated/sdk.gen.ts
var healthEndpointHealthGet = (options) => (options?.client ?? client).get({ url: "/health", ...options });
var getVersion = (options) => (options?.client ?? client).get({
  url: "/version",
  ...options
});
var metricsEndpointMetricsGet = (options) => (options?.client ?? client).get({ url: "/metrics", ...options });
var getGraph = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/graph", ...options });
var listMemories = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/memories/list", ...options });
var getMemory = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/memories/{memory_id}", ...options });
var getObservationHistory = (options) => (options.client ?? client).get({
  url: "/v1/default/banks/{bank_id}/memories/{memory_id}/history",
  ...options
});
var recallMemories = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/memories/recall",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var reflect = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/reflect",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var listBanks = (options) => (options?.client ?? client).get({ url: "/v1/default/banks", ...options });
var getAgentStats = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/stats", ...options });
var listEntities = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/entities", ...options });
var getEntity = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/entities/{entity_id}", ...options });
var regenerateEntityObservations = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/entities/{entity_id}/regenerate",
  ...options
});
var listMentalModels = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/mental-models", ...options });
var createMentalModel = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/mental-models",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var deleteMentalModel = (options) => (options.client ?? client).delete({
  url: "/v1/default/banks/{bank_id}/mental-models/{mental_model_id}",
  ...options
});
var getMentalModel = (options) => (options.client ?? client).get({
  url: "/v1/default/banks/{bank_id}/mental-models/{mental_model_id}",
  ...options
});
var updateMentalModel = (options) => (options.client ?? client).patch({
  url: "/v1/default/banks/{bank_id}/mental-models/{mental_model_id}",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var getMentalModelHistory = (options) => (options.client ?? client).get({
  url: "/v1/default/banks/{bank_id}/mental-models/{mental_model_id}/history",
  ...options
});
var refreshMentalModel = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/mental-models/{mental_model_id}/refresh",
  ...options
});
var listDirectives = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/directives", ...options });
var createDirective = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/directives",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var deleteDirective = (options) => (options.client ?? client).delete({
  url: "/v1/default/banks/{bank_id}/directives/{directive_id}",
  ...options
});
var getDirective = (options) => (options.client ?? client).get({
  url: "/v1/default/banks/{bank_id}/directives/{directive_id}",
  ...options
});
var updateDirective = (options) => (options.client ?? client).patch({
  url: "/v1/default/banks/{bank_id}/directives/{directive_id}",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var listDocuments = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/documents", ...options });
var deleteDocument = (options) => (options.client ?? client).delete({ url: "/v1/default/banks/{bank_id}/documents/{document_id}", ...options });
var getDocument = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/documents/{document_id}", ...options });
var updateDocument = (options) => (options.client ?? client).patch({
  url: "/v1/default/banks/{bank_id}/documents/{document_id}",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var listTags = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/tags", ...options });
var getChunk = (options) => (options.client ?? client).get({ url: "/v1/default/chunks/{chunk_id}", ...options });
var listOperations = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/operations", ...options });
var cancelOperation = (options) => (options.client ?? client).delete({
  url: "/v1/default/banks/{bank_id}/operations/{operation_id}",
  ...options
});
var getOperationStatus = (options) => (options.client ?? client).get({
  url: "/v1/default/banks/{bank_id}/operations/{operation_id}",
  ...options
});
var retryOperation = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/operations/{operation_id}/retry",
  ...options
});
var getBankProfile = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/profile", ...options });
var updateBankDisposition = (options) => (options.client ?? client).put({
  url: "/v1/default/banks/{bank_id}/profile",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var addBankBackground = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/background",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var deleteBank = (options) => (options.client ?? client).delete({ url: "/v1/default/banks/{bank_id}", ...options });
var updateBank = (options) => (options.client ?? client).patch({
  url: "/v1/default/banks/{bank_id}",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var createOrUpdateBank = (options) => (options.client ?? client).put({
  url: "/v1/default/banks/{bank_id}",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var clearObservations = (options) => (options.client ?? client).delete({ url: "/v1/default/banks/{bank_id}/observations", ...options });
var recoverConsolidation = (options) => (options.client ?? client).post({ url: "/v1/default/banks/{bank_id}/consolidation/recover", ...options });
var clearMemoryObservations = (options) => (options.client ?? client).delete({
  url: "/v1/default/banks/{bank_id}/memories/{memory_id}/observations",
  ...options
});
var resetBankConfig = (options) => (options.client ?? client).delete({ url: "/v1/default/banks/{bank_id}/config", ...options });
var getBankConfig = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/config", ...options });
var updateBankConfig = (options) => (options.client ?? client).patch({
  url: "/v1/default/banks/{bank_id}/config",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var triggerConsolidation = (options) => (options.client ?? client).post({ url: "/v1/default/banks/{bank_id}/consolidate", ...options });
var listWebhooks = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/webhooks", ...options });
var createWebhook = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/webhooks",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var deleteWebhook = (options) => (options.client ?? client).delete({ url: "/v1/default/banks/{bank_id}/webhooks/{webhook_id}", ...options });
var updateWebhook = (options) => (options.client ?? client).patch({
  url: "/v1/default/banks/{bank_id}/webhooks/{webhook_id}",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var listWebhookDeliveries = (options) => (options.client ?? client).get({
  url: "/v1/default/banks/{bank_id}/webhooks/{webhook_id}/deliveries",
  ...options
});
var clearBankMemories = (options) => (options.client ?? client).delete({ url: "/v1/default/banks/{bank_id}/memories", ...options });
var retainMemories = (options) => (options.client ?? client).post({
  url: "/v1/default/banks/{bank_id}/memories",
  ...options,
  headers: {
    "Content-Type": "application/json",
    ...options.headers
  }
});
var fileRetain = (options) => (options.client ?? client).post({
  ...formDataBodySerializer,
  url: "/v1/default/banks/{bank_id}/files/retain",
  ...options,
  headers: {
    "Content-Type": null,
    ...options.headers
  }
});
var listAuditLogs = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/audit-logs", ...options });
var auditLogStats = (options) => (options.client ?? client).get({ url: "/v1/default/banks/{bank_id}/audit-logs/stats", ...options });

// src/index.ts
var HindsightError = class extends Error {
  constructor(message, statusCode, details) {
    super(message);
    this.name = "HindsightError";
    this.statusCode = statusCode;
    this.details = details;
  }
};
var HindsightClient = class {
  constructor(options) {
    this.client = createClient(
      createConfig({
        baseUrl: options.baseUrl,
        headers: options.apiKey ? { Authorization: `Bearer ${options.apiKey}` } : void 0
      })
    );
  }
  /**
   * Validates the API response and throws an error if the request failed.
   */
  validateResponse(response, operation) {
    if (!response.data) {
      const error = response.error;
      const httpResponse = response.response;
      const statusCode = httpResponse?.status;
      const details = error?.detail || error?.message || error;
      throw new HindsightError(
        `${operation} failed: ${JSON.stringify(details)}`,
        statusCode,
        details
      );
    }
    return response.data;
  }
  /**
   * Retain a single memory for a bank.
   */
  async retain(bankId, content, options) {
    const item = { content };
    if (options?.timestamp) {
      item.timestamp = options.timestamp instanceof Date ? options.timestamp.toISOString() : options.timestamp;
    }
    if (options?.context) {
      item.context = options.context;
    }
    if (options?.metadata) {
      item.metadata = options.metadata;
    }
    if (options?.documentId) {
      item.document_id = options.documentId;
    }
    if (options?.entities) {
      item.entities = options.entities;
    }
    if (options?.tags) {
      item.tags = options.tags;
    }
    const response = await retainMemories({
      client: this.client,
      path: { bank_id: bankId },
      body: { items: [item], async: options?.async }
    });
    return this.validateResponse(response, "retain");
  }
  /**
   * Retain multiple memories in batch.
   */
  async retainBatch(bankId, items, options) {
    const processedItems = items.map((item) => ({
      content: item.content,
      context: item.context,
      metadata: item.metadata,
      document_id: item.document_id,
      entities: item.entities,
      tags: item.tags,
      observation_scopes: item.observation_scopes,
      strategy: item.strategy,
      timestamp: item.timestamp instanceof Date ? item.timestamp.toISOString() : item.timestamp
    }));
    const itemsWithDocId = processedItems.map((item) => ({
      ...item,
      document_id: item.document_id || options?.documentId
    }));
    const response = await retainMemories({
      client: this.client,
      path: { bank_id: bankId },
      body: {
        items: itemsWithDocId,
        document_tags: options?.documentTags,
        async: options?.async
      }
    });
    return this.validateResponse(response, "retainBatch");
  }
  /**
   * Upload files and retain their contents as memories.
   *
   * Files are automatically converted to text (PDF, DOCX, images via OCR, audio via
   * transcription, and more) and ingested as memories. Processing is always asynchronous —
   * use the returned operation IDs to track progress via the operations endpoint.
   *
   * @param bankId - The memory bank ID
   * @param files - Array of File or Blob objects to upload
   * @param options - Optional settings: context, documentTags, filesMetadata
   */
  async retainFiles(bankId, files, options) {
    const meta = options?.filesMetadata ?? files.map(() => options?.context ? { context: options.context } : {});
    const requestBody = JSON.stringify({
      files_metadata: meta
    });
    const response = await fileRetain({
      client: this.client,
      path: { bank_id: bankId },
      body: { files, request: requestBody }
    });
    return this.validateResponse(response, "retainFiles");
  }
  /**
   * Recall memories with a natural language query.
   */
  async recall(bankId, query, options) {
    const response = await recallMemories({
      client: this.client,
      path: { bank_id: bankId },
      body: {
        query,
        types: options?.types,
        max_tokens: options?.maxTokens,
        budget: options?.budget || "mid",
        trace: options?.trace,
        query_timestamp: options?.queryTimestamp,
        include: {
          entities: options?.includeEntities === false ? null : options?.includeEntities ? { max_tokens: options?.maxEntityTokens ?? 500 } : void 0,
          chunks: options?.includeChunks ? { max_tokens: options?.maxChunkTokens ?? 8192 } : void 0,
          source_facts: options?.includeSourceFacts ? { max_tokens: options?.maxSourceFactsTokens ?? 4096 } : void 0
        },
        tags: options?.tags,
        tags_match: options?.tagsMatch
      }
    });
    return this.validateResponse(response, "recall");
  }
  /**
   * Reflect and generate a contextual answer using the bank's identity and memories.
   */
  async reflect(bankId, query, options) {
    const response = await reflect({
      client: this.client,
      path: { bank_id: bankId },
      body: {
        query,
        context: options?.context,
        budget: options?.budget || "low",
        tags: options?.tags,
        tags_match: options?.tagsMatch
      }
    });
    return this.validateResponse(response, "reflect");
  }
  /**
   * List memories with pagination.
   */
  async listMemories(bankId, options) {
    const response = await listMemories({
      client: this.client,
      path: { bank_id: bankId },
      query: {
        limit: options?.limit,
        offset: options?.offset,
        type: options?.type,
        q: options?.q
      }
    });
    return this.validateResponse(response, "listMemories");
  }
  /**
   * Create or update a bank with disposition, missions, and operational configuration.
   */
  async createBank(bankId, options = {}) {
    const response = await createOrUpdateBank({
      client: this.client,
      path: { bank_id: bankId },
      body: {
        name: options.name,
        mission: options.mission,
        reflect_mission: options.reflectMission,
        background: options.background,
        disposition: options.disposition,
        disposition_skepticism: options.dispositionSkepticism,
        disposition_literalism: options.dispositionLiteralism,
        disposition_empathy: options.dispositionEmpathy,
        retain_mission: options.retainMission,
        retain_extraction_mode: options.retainExtractionMode,
        retain_custom_instructions: options.retainCustomInstructions,
        retain_chunk_size: options.retainChunkSize,
        enable_observations: options.enableObservations,
        observations_mission: options.observationsMission
      }
    });
    return this.validateResponse(response, "createBank");
  }
  /**
   * Set or update the reflect mission for a memory bank.
   * @deprecated Use createBank({ reflectMission: '...' }) instead.
   */
  async setMission(bankId, mission) {
    return this.createBank(bankId, { reflectMission: mission });
  }
  /**
   * Get a bank's profile.
   */
  async getBankProfile(bankId) {
    const response = await getBankProfile({
      client: this.client,
      path: { bank_id: bankId }
    });
    return this.validateResponse(response, "getBankProfile");
  }
  /**
   * Get the resolved configuration for a bank, including any bank-level overrides.
   *
   * Can be disabled on the server by setting `HINDSIGHT_API_ENABLE_BANK_CONFIG_API=false`.
   */
  async getBankConfig(bankId) {
    const response = await getBankConfig({
      client: this.client,
      path: { bank_id: bankId }
    });
    return this.validateResponse(response, "getBankConfig");
  }
  /**
   * Update configuration overrides for a bank.
   *
   * Can be disabled on the server by setting `HINDSIGHT_API_ENABLE_BANK_CONFIG_API=false`.
   *
   * @param bankId - The memory bank ID
   * @param options - Fields to override
   */
  async updateBankConfig(bankId, options) {
    const updates = {};
    if (options.reflectMission !== void 0) updates.reflect_mission = options.reflectMission;
    if (options.retainMission !== void 0) updates.retain_mission = options.retainMission;
    if (options.retainExtractionMode !== void 0) updates.retain_extraction_mode = options.retainExtractionMode;
    if (options.retainCustomInstructions !== void 0)
      updates.retain_custom_instructions = options.retainCustomInstructions;
    if (options.retainChunkSize !== void 0) updates.retain_chunk_size = options.retainChunkSize;
    if (options.enableObservations !== void 0) updates.enable_observations = options.enableObservations;
    if (options.observationsMission !== void 0) updates.observations_mission = options.observationsMission;
    if (options.dispositionSkepticism !== void 0) updates.disposition_skepticism = options.dispositionSkepticism;
    if (options.dispositionLiteralism !== void 0) updates.disposition_literalism = options.dispositionLiteralism;
    if (options.dispositionEmpathy !== void 0) updates.disposition_empathy = options.dispositionEmpathy;
    const response = await updateBankConfig({
      client: this.client,
      path: { bank_id: bankId },
      body: { updates }
    });
    return this.validateResponse(response, "updateBankConfig");
  }
  /**
   * Reset all bank-level configuration overrides, reverting to server defaults.
   *
   * Can be disabled on the server by setting `HINDSIGHT_API_ENABLE_BANK_CONFIG_API=false`.
   */
  async resetBankConfig(bankId) {
    const response = await resetBankConfig({
      client: this.client,
      path: { bank_id: bankId }
    });
    return this.validateResponse(response, "resetBankConfig");
  }
  /**
   * Delete a bank.
   */
  async deleteBank(bankId) {
    const response = await deleteBank({
      client: this.client,
      path: { bank_id: bankId }
    });
    if (response.error) {
      throw new Error(`deleteBank failed: ${JSON.stringify(response.error)}`);
    }
  }
  // Directive methods
  /**
   * Create a directive (hard rule for reflect).
   */
  async createDirective(bankId, name, content, options) {
    const response = await createDirective({
      client: this.client,
      path: { bank_id: bankId },
      body: {
        name,
        content,
        priority: options?.priority ?? 0,
        is_active: options?.isActive ?? true,
        tags: options?.tags
      }
    });
    return this.validateResponse(response, "createDirective");
  }
  /**
   * List all directives in a bank.
   */
  async listDirectives(bankId, options) {
    const response = await listDirectives({
      client: this.client,
      path: { bank_id: bankId },
      query: { tags: options?.tags }
    });
    return this.validateResponse(response, "listDirectives");
  }
  /**
   * Get a specific directive.
   */
  async getDirective(bankId, directiveId) {
    const response = await getDirective({
      client: this.client,
      path: { bank_id: bankId, directive_id: directiveId }
    });
    return this.validateResponse(response, "getDirective");
  }
  /**
   * Update a directive.
   */
  async updateDirective(bankId, directiveId, options) {
    const response = await updateDirective({
      client: this.client,
      path: { bank_id: bankId, directive_id: directiveId },
      body: {
        name: options.name,
        content: options.content,
        priority: options.priority,
        is_active: options.isActive,
        tags: options.tags
      }
    });
    return this.validateResponse(response, "updateDirective");
  }
  /**
   * Delete a directive.
   */
  async deleteDirective(bankId, directiveId) {
    const response = await deleteDirective({
      client: this.client,
      path: { bank_id: bankId, directive_id: directiveId }
    });
    if (response.error) {
      throw new Error(`deleteDirective failed: ${JSON.stringify(response.error)}`);
    }
  }
  // Mental Model methods
  /**
   * Create a mental model (runs reflect in background).
   */
  async createMentalModel(bankId, name, sourceQuery, options) {
    const response = await createMentalModel({
      client: this.client,
      path: { bank_id: bankId },
      body: {
        id: options?.id,
        name,
        source_query: sourceQuery,
        tags: options?.tags,
        max_tokens: options?.maxTokens,
        trigger: options?.trigger ? { refresh_after_consolidation: options.trigger.refreshAfterConsolidation } : void 0
      }
    });
    return this.validateResponse(response, "createMentalModel");
  }
  /**
   * List all mental models in a bank.
   */
  async listMentalModels(bankId, options) {
    const response = await listMentalModels({
      client: this.client,
      path: { bank_id: bankId },
      query: { tags: options?.tags }
    });
    return this.validateResponse(response, "listMentalModels");
  }
  /**
   * Get a specific mental model.
   */
  async getMentalModel(bankId, mentalModelId) {
    const response = await getMentalModel({
      client: this.client,
      path: { bank_id: bankId, mental_model_id: mentalModelId }
    });
    return this.validateResponse(response, "getMentalModel");
  }
  /**
   * Refresh a mental model to update with current knowledge.
   */
  async refreshMentalModel(bankId, mentalModelId) {
    const response = await refreshMentalModel({
      client: this.client,
      path: { bank_id: bankId, mental_model_id: mentalModelId }
    });
    return this.validateResponse(response, "refreshMentalModel");
  }
  /**
   * Update a mental model's metadata.
   */
  async updateMentalModel(bankId, mentalModelId, options) {
    const response = await updateMentalModel({
      client: this.client,
      path: { bank_id: bankId, mental_model_id: mentalModelId },
      body: {
        name: options.name,
        source_query: options.sourceQuery,
        tags: options.tags,
        max_tokens: options.maxTokens,
        trigger: options.trigger ? { refresh_after_consolidation: options.trigger.refreshAfterConsolidation } : void 0
      }
    });
    return this.validateResponse(response, "updateMentalModel");
  }
  /**
   * Delete a mental model.
   */
  async deleteMentalModel(bankId, mentalModelId) {
    const response = await deleteMentalModel({
      client: this.client,
      path: { bank_id: bankId, mental_model_id: mentalModelId }
    });
    if (response.error) {
      throw new Error(`deleteMentalModel failed: ${JSON.stringify(response.error)}`);
    }
  }
  /**
   * Get the change history of a mental model.
   */
  async getMentalModelHistory(bankId, mentalModelId) {
    const response = await getMentalModelHistory({
      client: this.client,
      path: { bank_id: bankId, mental_model_id: mentalModelId }
    });
    return this.validateResponse(response, "getMentalModelHistory");
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  HindsightClient,
  HindsightError,
  createClient,
  createConfig,
  sdk
});
